controllers/storageclaim_controller.go

```diff
@@ -613,3 +643,103 @@ func (r *StorageClassClaimReconciler) createOrReplaceVolumeSnapshotClass(volumeS
 	}
 	return nil
 }
+
+func (r *StorageClaimReconciler) migrateStorageClassClaims() error {
+	storageClassClaimsList := &v1alpha1.StorageClassClaimList{}
+	if err := r.list(storageClassClaimsList); err != nil {
+		return fmt.Errorf("failed to list storageclassclaims")
+	}
+
+	for idx := range storageClassClaimsList.Items {
+		storageClassClaim := &storageClassClaimsList.Items[idx]
+		storageClaim := &v1alpha1.StorageClaim{}
+		storageClaim.Name = storageClassClaim.Name
+
+		// migrate storageclassclaim to storageclaim
+		_, err := controllerutil.CreateOrUpdate(r.ctx, r.Client, storageClaim, func() error {
+			storageClaim.Spec.Type = storageClassClaim.Spec.Type
+			storageClaim.Spec.EncryptionMethod = storageClassClaim.Spec.EncryptionMethod
+			storageClaim.Spec.StorageProfile = storageClassClaim.Spec.StorageProfile
+			storageClaim.Spec.StorageClient = storageClassClaim.Spec.StorageClient.DeepCopy()
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to copy spec from storageclassclaims to storageclaims: %v", err)
+		}
+
+		for idx := range storageClassClaim.Status.SecretNames {
+			secretName := storageClassClaim.Status.SecretNames[idx]
+			secret := &corev1.Secret{}
+			secret.Name = secretName
+			secret.Namespace = storageClassClaim.Spec.StorageClient.Namespace
+			if err = r.get(secret); err != nil {
+				return fmt.Errorf("failed to get secret %s: %v", client.ObjectKeyFromObject(secret), err)
+			}
+
+			if err = controllerutil.SetOwnerReference(storageClaim, secret, r.Scheme); err != nil {
+				return fmt.Errorf("failed to set owner reference on secret %s: %v", client.ObjectKeyFromObject(secret), err)
+			}
+		}
+
+		// remove finalizer on existing storageclassclaim
+		finalizerUpdated := controllerutil.RemoveFinalizer(storageClassClaim, storageClassClaimFinalizer)
+		if finalizerUpdated {
+			if err := r.update(storageClassClaim); err != nil {
+				return fmt.Errorf("failed to remove finalizer on storageclassclaim %q: %v", storageClassClaim.Name, err)
+			}
+		}
+
+		// migration is successful delete the storageclassclaim
+		if err := r.delete(storageClassClaim); err != nil {
+			return fmt.Errorf("failed to delete storageclassclaim %q: %v", storageClassClaim.Name, err)
+		}
+
+		r.log.Info(fmt.Sprintf("Successfully migrated storageclassclaim %q to storageclass %q", storageClassClaim.Name, storageClaim.Name))
+	}
```

Comment:
Preferably migrate one by one:
- Create new claim
- copy data
- remove finalizer on old claim
- delete old claim 
- move to next claim in the list.
 
 This way if you fail for a specific claim you will not repeat completed steps on claims that already got migrated.

Reply:
Sorry. I read the code incorrectly. You are right.